File: CSC310-Python/Module00-Course-Introduction-and-Data-Structures-Review/debug01/two_stack_queue.py
"""A FIFO queue built from two LIFO stacks.

Elements are pushed onto an inbox stack.  When a dequeue is requested,
elements are moved to an outbox stack so that the oldest element is on
top.  Each element is moved at most once, so the amortized cost of every
operation is O(1).
"""


class TwoStackQueue:
    def __init__(self):
        self._inbox = []
        self._outbox = []

    def __len__(self):
        return len(self._inbox) + len(self._outbox)

    def is_empty(self):
        return len(self) == 0

    def enqueue(self, item):
        self._inbox.append(item)

    def _shift(self):
        # Move everything from the inbox to the outbox, reversing order.
        while self._inbox:
            self._outbox.append(self._inbox.pop())

    def dequeue(self):
        if self.is_empty():
            raise IndexError("dequeue from empty queue")
        self._shift()
        return self._outbox.pop()

    def peek(self):
        if self.is_empty():
            raise IndexError("peek at empty queue")
        self._shift()
        return self._outbox[-1]


def main():
    # Simulate a print server: jobs arrive in bursts and are printed in
    # arrival order.
    q = TwoStackQueue()
    served = []

    for job in ["report.pdf", "photo.png", "notes.txt"]:
        q.enqueue(job)
    served.append(q.dequeue())

    for job in ["slides.pptx", "invoice.pdf"]:
        q.enqueue(job)
    served.append(q.dequeue())
    served.append(q.dequeue())

    q.enqueue("resume.docx")
    while not q.is_empty():
        served.append(q.dequeue())

    print("Jobs printed, in order:")
    for i, job in enumerate(served, 1):
        print(f"  {i}. {job}")

    # A second check with integers, comparing against a plain list.
    q2 = TwoStackQueue()
    expected = []
    out = []
    for round_no in range(1, 5):
        for k in range(round_no):
            q2.enqueue(round_no * 10 + k)
            expected.append(round_no * 10 + k)
        out.append(q2.dequeue())
    while not q2.is_empty():
        out.append(q2.dequeue())
    print("Integer test matches FIFO order:", out == expected)
    print("Dequeued:", out)


if __name__ == "__main__":
    main()

File: CSC310-Cpp/_solutions/Module00-Course-Introduction-and-Data-Structures-Review/debug01/two_stack_queue.cpp
// two_stack_queue.cpp
// A FIFO queue built from two LIFO stacks (an "inbox" and an "outbox").
// Enqueue pushes onto the inbox; dequeue pops from the outbox, refilling
// the outbox from the inbox when necessary. Amortized O(1) per operation.
#include <iostream>
#include <stack>
#include <stdexcept>

class TwoStackQueue {
public:
    void enqueue(int value) {
        inbox_.push(value);
    }

    int dequeue() {
        shift();
        if (outbox_.empty()) {
            throw std::runtime_error("dequeue on empty queue");
        }
        int front = outbox_.top();
        outbox_.pop();
        return front;
    }

    int peek() {
        shift();
        if (outbox_.empty()) {
            throw std::runtime_error("peek on empty queue");
        }
        return outbox_.top();
    }

    bool empty() const {
        return inbox_.empty() && outbox_.empty();
    }

    std::size_t size() const {
        return inbox_.size() + outbox_.size();
    }

private:
    // Move everything from the inbox to the outbox so that the oldest
    // element ends up on top of the outbox. Only done when the outbox is
    // empty; otherwise the outbox already holds older elements.
    void shift() {
        if (!outbox_.empty()) {
            return;
        }
        while (!inbox_.empty()) {
            outbox_.push(inbox_.top());
            inbox_.pop();
        }
    }

    std::stack<int> inbox_;
    std::stack<int> outbox_;
};

int main() {
    TwoStackQueue q;

    q.enqueue(1);
    q.enqueue(2);
    q.enqueue(3);
    std::cout << "dequeue -> " << q.dequeue() << '\n';

    q.enqueue(4);
    q.enqueue(5);
    std::cout << "peek    -> " << q.peek() << '\n';

    std::cout << "draining: ";
    while (!q.empty()) {
        std::cout << q.dequeue() << ' ';
    }
    std::cout << '\n';

    // A second round to make sure the queue is reusable.
    for (int v = 10; v <= 14; ++v) {
        q.enqueue(v);
        if (v % 2 == 0) {
            std::cout << "dequeue -> " << q.dequeue() << '\n';
        }
    }
    std::cout << "remaining size: " << q.size() << '\n';
    std::cout << "draining: ";
    while (!q.empty()) {
        std::cout << q.dequeue() << ' ';
    }
    std::cout << '\n';
    return 0;
}
